File: backend/app/schemas/trade.py
from datetime import date, datetime, time
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field, field_serializer
from typing import Literal


class TradeCreate(BaseModel):
    stock_symbol: str
    trade_type: str
    quantity: int
    price: Decimal
    trade_date: date
    broker: str | None = None
    import_source: str | None = None
    emotion_tag: str | None = None
    notes: str | None = None
    screenshot_url: str | None = None
    entry_method: str | None = None
    trade_time: time | None = None
    instrument_type: str | None = None


class TradeResponse(BaseModel):
    id: int
    user_id: int
    stock_symbol: str
    trade_type: str
    quantity: int
    price: Decimal
    trade_date: date
    broker: str | None
    import_source: str | None
    emotion_tag: str | None
    notes: str | None
    screenshot_url: str | None
    entry_method: str | None
    trade_time: time | None
    instrument_type: str | None
    created_at: datetime

    model_config = {"from_attributes": True}


class TradeImportRequest(BaseModel):
    email_content: str = Field(..., min_length=1)


class AutoCapturedTradeInput(BaseModel):
    stock_symbol: str
    trade_type: Literal["BUY", "SELL"]
    quantity: int = Field(..., gt=0)
    price: Decimal = Field(..., gt=0)
    trade_date: date
    trade_time: time | None = None
    instrument_type: str | None = None
    entry_method: str | None = None
    emotion_tag: str | None = None
    notes: str | None = None
    screenshot_url: str | None = None


class AutoCaptureRequest(BaseModel):
    broker: Literal["zerodha", "groww"]
    capture_method: Literal["dom"]
    trades: list[AutoCapturedTradeInput] = Field(default_factory=list)


class TradeAnnotationUpdateRequest(BaseModel):
    emotion_tag: str | None = None
    note: str | None = None


class TradeImportResponse(BaseModel):
    imported: int
    trades: list[TradeResponse] = Field(default_factory=list)
    imported_count: int = 0
    duplicate_count: int = 0
    imported_trade_ids: list[int] = Field(default_factory=list)
    detected_broker: str | None = None
    mode: Literal["imported", "manual_mapping_required"] = "imported"
    preview_headers: list[str] = Field(default_factory=list)
    preview_rows: list[dict[str, str]] = Field(default_factory=list)
    message: str | None = None


class TradesSummary(BaseModel):
    total_trades: int
    total_invested: Decimal
    unique_symbols: int
    net_pnl_today: Decimal = Decimal("0")
    max_loss_threshold: Decimal = Decimal("0")
    hidden_trade_count: int = 0

    @field_serializer("total_invested", "net_pnl_today", "max_loss_threshold", when_used="json")
    def serialize_decimal_fields(self, value: Decimal) -> float:
        return float(value)


class CompletedTradeResponse(BaseModel):
    id: int
    user_id: int
    stock_symbol: str
    entry_date: date
    exit_date: date
    entry_price: Decimal
    exit_price: Decimal
    quantity: int
    pnl: Decimal
    total_charges: Decimal
    net_pnl: Decimal
    return_pct: Decimal
    holding_days: int
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)

    @field_serializer(
        "entry_price",
        "exit_price",
        "pnl",
        "total_charges",
        "net_pnl",
        "return_pct",
        when_used="json",
    )
    def serialize_decimal_fields(self, value: Decimal) -> float:
        return float(value)


class PaginatedTradesResponse(BaseModel):
    trades: list[TradeResponse]
    total: int
    hidden_trade_count: int
    is_limited: bool


class PaginatedCompletedTradesResponse(BaseModel):
    trades: list[CompletedTradeResponse]
    total: int
    hidden_trade_count: int
    is_limited: bool

File: frontend/src/types/trade.ts
export interface Trade {
  id: number;
  user_id: number;
  stock_symbol: string;
  trade_type: string;
  quantity: number;
  price: number;
  trade_date: string;
  broker: string | null;
  import_source: string | null;
  emotion_tag?: string | null;
  notes?: string | null;
  screenshot_url?: string | null;
  entry_method?: string | null;
  trade_time?: string | null;
  instrument_type?: string | null;
  created_at: string;
}

export interface TradeImportResponse {
  imported: number;
  imported_count: number;
  duplicate_count: number;
  trades: Trade[];
  imported_trade_ids?: number[];
  detected_broker?: string | null;
  mode?: "imported" | "manual_mapping_required";
  preview_headers: string[];
  preview_rows: Record<string, string>[];
  message?: string | null;
}

export interface TradesSummary {
  total_trades: number;
  total_invested: number;
  unique_symbols: number;
  net_pnl_today: number;
  max_loss_threshold: number;
}

export interface CompletedTrade {
  id: number;
  user_id: number;
  stock_symbol: string;
  entry_date: string;
  exit_date: string;
  entry_price: number;
  exit_price: number;
  quantity: number;
  pnl: number;
  total_charges: number;
  net_pnl: number;
  return_pct: number;
  holding_days: number;
  created_at: string;
}

export interface TradeSetup {
  id: number;
  user_id: number;
  symbol: string | null;
  thesis: string | null;
  entry_price: number | null;
  stop_loss_price: number | null;
  target_price: number | null;
  target2_price: number | null;
  conviction_score: number | null;
  checklist_responses: Record<string, unknown> | null;
  position_size: number | null;
  risk_amount: number | null;
  risk_score: number | null;
  risk_level: string | null;
  linked_trade_id: number | null;
  linked_at: string | null;
  created_at: string;
}
